gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/DocumentReferenceMapper.java

```diff
@@ -0,0 +1,218 @@
+        documentReference.getMeta().addProfile(META_PROFILE);
+        documentReference.setStatus(Enumerations.DocumentReferenceStatus.CURRENT);
+        documentReference.setId(id);
+        documentReference.setType(getType(narrativeStatement));
+        documentReference.setSubject(new Reference(patient));
+        documentReference.setIndexedElement(getIndexed(ehrExtract));
+        documentReference.setAuthor(getAuthor(narrativeStatement, ehrComposition));
+        documentReference.setDescription(buildDescription(narrativeStatement));
+
+        if (narrativeStatement.getAvailabilityTime() != null && narrativeStatement.getAvailabilityTime().getValue() != null) {
+            documentReference.setCreatedElement(
+                DateFormatUtil.parseToDateTimeType(narrativeStatement.getAvailabilityTime().getValue()));
+        }
+
+        var encounterReference =
+            EncounterReferenceUtil.getEncounterReference(List.of(ehrComposition), encounterList, ehrComposition.getId().getRoot());
+
+        if (encounterReference != null) {
+            DocumentReference.DocumentReferenceContextComponent documentReferenceContextComponent
+                = new DocumentReference.DocumentReferenceContextComponent().setEncounter(encounterReference);
+
+            documentReference.setContext(documentReferenceContextComponent);
+        }
+
+        setContentAttachments(documentReference, narrativeStatement);
+
+        return documentReference;
+    }
+
+    private List<RCMRMT030101UK04NarrativeStatement> getNarrativeStatements(RCMRMT030101UK04EhrComposition ehrComposition) {
+        return ehrComposition.getComponent()
+            .stream()
+            .map(RCMRMT030101UK04Component4::getNarrativeStatement)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+    }
+
+    private CodeableConcept getType(RCMRMT030101UK04NarrativeStatement narrativeStatement) {
+        var referenceToExternalDocument = narrativeStatement.getReference().get(0).getReferredToExternalDocument();
+
+        if (referenceToExternalDocument != null && referenceToExternalDocument.getCode() != null) {
+            if (referenceToExternalDocument.getCode().getOriginalText() == null
+                && referenceToExternalDocument.getCode().getDisplayName() != null) {
+                return codeableConceptMapper.mapToCodeableConcept(referenceToExternalDocument.getCode())
+                    .setText(referenceToExternalDocument.getCode().getDisplayName());
+            } else if (referenceToExternalDocument.getCode().getOriginalText() != null) {
+                return codeableConceptMapper.mapToCodeableConcept(referenceToExternalDocument.getCode())
+                    .setText(referenceToExternalDocument.getCode().getOriginalText());
+            }
+        }
+
+        return null;
+    }
+
+    private InstantType getIndexed(RCMRMT030101UK04EhrExtract ehrExtract) {
+        if (ehrExtract.getAuthor() != null) {
```

Comment:
all of these get classes for the likes of author, time, availabilityTime etc may have 'has' functions which might look a bit cleaner than these != null or ==, may be worth looking into